c10/detail/InlineDeviceGuard.h

```diff
@@ -0,0 +1,137 @@
+#pragma once
+
+namespace c10 {
+namespace detail {
+
+/**
+ * InlineDeviceGuard is a helper class for implementing device type
+ * specialized DeviceGuards, e.g., CUDAGuard.  The key idea is that
+ * it is templated over DeviceGuardImpl, allowing it to devirtualize
+ * all calls.  The intention is that InlineDeviceGuard<CUDAGuardImpl>
+ * should be *as* efficient as straight line code that calls
+ * cudaGetDevice-cudaSetDevice.  InlineDeviceGuard can only be used
+ * from code that links against the relevant backend library, e.g.,
+ * CUDA library.
+ *
+ * Users of InlineDeviceGuard may also find that the interface is also
+ * more expressive.  This is because DeviceGuard cannot handle
+ * set_device() in a reasonable way (consider what happens if you
+ * set_device for CUDA, and then set_device for HIP!)  Because
+ * InlineDeviceGuard is specialized for a specific device type,
+ * it doesn't have to worry about this API situation.  (In principle,
+ * you might find a use for this interface even without devirtualization;
+ * but a better solution in such cases is to move your code into a compilation
+ * unit that links against, e.g., CUDA).
+ *
+ * This code assumes that there are no non-bracketed calls to the underlying
+ * setDevice() within the body of the guard.
+ */
+template <typename T>
+class InlineDeviceGuard {
+public:
+  /// Default constructor, reads the current device so that
+  /// we may reset the device to the current device on destruction.
+  explicit InlineDeviceGuard() {
+    initialize();
+  }
+
+  /// Set the current device to the passed Device.
+  explicit InlineDeviceGuard(Device device) {
+    set_device(device); // In Optimizer We Trust
```

Comment:
We need an initialize() here as well, otherwise current_device_  will not be set.

Reply:
Thanks. I got this (and others) after writing a full set of tests for the DeviceGuard :)